src/coreclr/jit/lowerarmarch.cpp

```diff
@@ -2266,156 +2270,105 @@ void Lowering::ContainCheckCompare(GenTreeOp* cmp)
     {
-        // Already have a contained chain.
-        return true;
+        return nullptr;
     }
-    // Can the child be contained.
-    else if (IsInvariantInRange(child, parent))
+
+    GenCondition cond1;
+    if (!TryLowerConditionToFlagsNode(tree, op1, &cond1))
     {
-        if (child->OperIs(GT_AND))
-        {
-            // If Op2 is not contained, then try to contain it.
-            if (!child->isContainedCompareChainSegment(child->AsOp()->gtGetOp2()))
-            {
-                if (!ContainCheckCompareChain(child->gtGetOp2(), child, startOfChain))
-                {
-                    // Op2 must be contained in order to contain Op1 or the AND.
-                    return false;
-                }
-            }
+        return nullptr;
+    }
 
-            // If Op1 is not contained, then try to contain it.
-            if (!child->isContainedCompareChainSegment(child->AsOp()->gtGetOp1()))
-            {
-                if (!ContainCheckCompareChain(child->gtGetOp1(), child, startOfChain))
-                {
-                    return false;
-                }
-            }
+    BlockRange().Remove(op2);
+    BlockRange().InsertBefore(tree, op2);
 
-            // Contain the AND.
-            child->SetContained();
-            return true;
-        }
-        else if (child->OperIsCmpCompare())
-        {
-            child->AsOp()->SetContained();
+    GenCondition cond2 = GenCondition::FromRelop(op2);
+    op2->SetOper(GT_CCMP);
+    op2->gtType = TYP_VOID;
+    op2->gtFlags |= GTF_SET_FLAGS;
 
-            // Ensure the children of the compare are contained correctly.
-            child->AsOp()->gtGetOp1()->ClearContained();
-            child->AsOp()->gtGetOp2()->ClearContained();
-            ContainCheckConditionalCompare(child->AsOp());
-            *startOfChain = child;
-            return true;
-        }
-    }
+    op2->gtGetOp1()->ClearContained();
+    op2->gtGetOp2()->ClearContained();
 
-    return false;
+    GenTreeCCMP* ccmp = op2->AsCCMP();
+    ccmp->gtCondition = cond1;
+    // If the first comparison fails, set the condition flags to something that
+    // makes the second one fail.
+    ccmp->gtFlagsVal = TruthifyingFlags(GenCondition::Reverse(cond2));
+    ContainCheckConditionalCompare(ccmp);
+
+    tree->SetOper(GT_SETCC);
+    tree->AsCC()->gtCondition = cond2;
+
+    return tree->gtNext;
 }
 
 //------------------------------------------------------------------------
-// ContainCheckCompareChainForAnd : Determine if an AND is a containable chain
+// TruthifyingFlags: Get a flags immediate that will make a specified condition true.
 //
 // Arguments:
-//    node - pointer to the node
+//    condition - the condition.
 //
-void Lowering::ContainCheckCompareChainForAnd(GenTree* tree)
+// Returns:
+//    A flags immediate that, if those flags were set, would cause the specified condition to be true.
+//
+insCflags Lowering::TruthifyingFlags(GenCondition condition)
```

Comment:
Can we repurpose `InsCflagsForCcmp` ? Move it to a GenTree, make it static and use it from codegen as well as lower?

Reply:
We don't need that function in codegen anymore -- this PR is removing it from codegenarm64.cpp (it is effectively the same function, except without reversing the condition). In a follow-up I will add the ability to generate ccmp for `OR(relop, relop)` too which will use that function again.

Reply:
Ah, for some reason I didn't noticed that you have removed `InsCflagsForCcmp()`.